blog/nse_portfolio_sorts/index.qmd

```diff
@@ -0,0 +1,931 @@
+
+```{r}
+execute_sorts <- function(sorting_variable, drop_smallNYSE_at, include_financials, include_utilities, drop_bookequity,
+                          drop_earnings, drop_stock_age_at, drop_price_at, sv_lag, formation_time, n_portfolios_main,
+                          sorting_method, n_portfolios_secondary, exchanges, value_weighted) {
+  # Select data
+  data_sorts <- handle_data(include_financials = include_financials, 
+                            include_utilities = include_utilities, 
+                            drop_smallNYSE_at = drop_smallNYSE_at,  
+                            drop_price_at = drop_price_at,
+                            drop_stock_age_at = drop_stock_age_at,
+                            drop_earnings = drop_earnings, 
+                            drop_bookequity = drop_bookequity,
+                            sv_lag = sv_lag)
+  
+  # Rebalancing
+  ## Monthly
+  if(formation_time == "monthly") {
+    data_return <- rebalance_mon(data = data_sorts, 
+                                 sorting_variable = sorting_variable, 
+                                 sorting_method = sorting_method, 
+                                 n_portfolios_main = n_portfolios_main, 
+                                 n_portfolios_secondary = n_portfolios_secondary, 
+                                 exchanges = exchanges, 
+                                 value_weighted = value_weighted)
+  }
+  
+  ## Monthly
+  if(formation_time == "FF") {
+    data_return <- rebalance_ann(data = data_sorts, 
+                                 sorting_variable = sorting_variable, 
+                                 sorting_method = sorting_method, 
+                                 n_portfolios_main = n_portfolios_main, 
+                                 n_portfolios_secondary = n_portfolios_secondary, 
+                                 exchanges = exchanges, 
+                                 value_weighted = value_weighted)
+  }
+  
+  # Compute return differential
+  data_return |> 
+    group_by(month) |> 
+    summarize(premium = ret[portfolio == max(portfolio)] - ret[portfolio == min(portfolio)],
+              .groups = "drop") |> 
+    pull(premium) |> 
+    mean() * -100
+}
+```
+
+
+## Applying the functions
+
+Finally, we have data, decisions, and functions. Indeed, we are now ready to implement the portfolio sort, right? Yes! Just let me briefly discuss how the implementation works.
+
+We have a grid with 69,120 specifications. For each of these specifications, we want to estimate a return differential. This is most easily achieved with a `pmap()` call. However, we want to parallelize the operation to leverage the multiple cores of our device. Hence, we have to use the package `furrr` and a `future_pmap()` instead. As a side note, I also have to increase my maximum memory for each worker, which I do with the call of `options()`. 
+
+```{r}
+#| eval: false
+library(furrr)
+
+options(future.globals.maxSize= 891289600)
```

Comment:
Ja, hier ist auch da Performance Issue begraben: jeder job braucht einfach alle Daten in memory, daher müssen auch jeder worker alle Daten haben. Ich vermute, dass man mit dtplyr schon noch etwas an Speed gewinnen kann, da dann die filters schneller funktionieren. Ansonsten hab ich ad hoc leider keinen quick fix parat